gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -0,0 +1,502 @@
+                });
+            }
+        };
+
+        // TODO: Find a way to avoid looping/polling to know when a payment is
+        // completed. `ldk-node` provides `PaymentSuccessful` and `PaymentFailed`
+        // events, but interacting with the node event queue here isn't
+        // straightforward.
+        loop {
+            if let Some(payment_details) = self.node.payment(&payment_id) {
+                match payment_details.status {
+                    PaymentStatus::Pending => {}
+                    PaymentStatus::Succeeded => {
+                        if let PaymentKind::Bolt11 {
+                            preimage: Some(preimage),
+                            ..
+                        } = payment_details.kind
+                        {
+                            return Ok(PayInvoiceResponse {
+                                preimage: preimage.0.to_vec(),
+                            });
+                        }
+                    }
+                    PaymentStatus::Failed => {
+                        return Err(LightningRpcError::FailedPayment {
+                            failure_reason: "LDK payment failed".to_string(),
+                        });
+                    }
+                }
+            }
+            fedimint_core::runtime::sleep(Duration::from_millis(100)).await;
+        }
+    }
+
+    async fn route_htlcs<'a>(
+        mut self: Box<Self>,
+        _task_group: &TaskGroup,
+    ) -> Result<(RouteHtlcStream<'a>, Arc<dyn ILnRpcClient>), LightningRpcError> {
+        let route_htlc_stream = match self.htlc_stream_receiver_or.take() {
+            Some(stream) => Ok(Box::pin(ReceiverStream::new(stream))),
+            None => Err(LightningRpcError::FailedToRouteHtlcs {
+                failure_reason:
+                    "Stream does not exist. Likely was already taken by calling `route_htlcs()`."
+                        .to_string(),
+            }),
+        }?;
+
+        Ok((route_htlc_stream, Arc::new(*self)))
+    }
+
+    async fn complete_htlc(
+        &self,
+        htlc: InterceptHtlcResponse,
+    ) -> Result<EmptyResponse, LightningRpcError> {
+        let InterceptHtlcResponse {
+            action,
+            payment_hash,
+            incoming_chan_id: _,
+            htlc_id: _,
+        } = htlc;
+
+        let ph = PaymentHash(payment_hash.clone().try_into().map_err(|_| {
+            LightningRpcError::FailedToCompleteHtlc {
+                failure_reason: "Failed to parse payment hash".to_string(),
+            }
+        })?);
+
+        // TODO: Get the actual amount from the LDK node. Probably makes the
+        // most sense to pipe it through the `InterceptHtlcResponse` struct.
+        // This value is only used by `ldk-node` to ensure that the amount
+        // claimed isn't less than the amount expected, but we've already
+        // verified that the amount is correct when we intercepted the HTLC.
```

Comment:
nit: "intercepted the payment" not HTLC. Once LNv2 is merged we should refactor all HTLC names out of the codebase